pkg/impersonation/impersonation.go

```diff
@@ -373,12 +374,23 @@ func (i *Impersonator) getUser(userInfo user.Info) (user.Info, error) {
 				}
 			}
 		}
+		for _, exs := range attribs.ExtraByProvider {
```

Comment:
I think I understand the point of this code, but I have a few questions on the functionality:
1) Why do we single out the "username" and "principalid" extras? I get that these are important extras, but the prior version of the code seems to have captured all extra values. (L380 in prior version)
2) I do see a couple of places where we get the extras from the UserInfo directly. Is this the "canonical way" that we would want this to happen?
3) Instead of replacing all of the extras from the userInfo would we instead want to merge the two maps to get all the extras possible (although I'm kind of thinking that ideally we would want these values to be equal, so really it shouldn't matter)?

Reply:
1. `username` and `principalid` are sent with the user's request and included in the impersonation headers, they are not chosen by us. An example would look something like this: `curl -H "Impersonate-User: user-abcde" -H "Impersonate-Group: cattle:authenticated" -H "Impersonate-Extra-PrincipalID: "github://abc" -H "Impersonate-Extra-Username: cmurphy"`, and if the ClusterRole does not include rules for those extra fields, the request will be denied due to missing RBAC rules.
2. This is the crux of the issue, the extras may not be included in the UserInfo object, depending on whether this is called from a role template binding controller or from the proxy request handler. It's more consistent to get the data from the UserAttribute object than to trust the UserInfo object.
3. With this change, if any extras are included in UserInfo they would be redundant with what is stored in the UserAttribute object. If that were not true, then the rules in the ClusterRole would flap depending on whether the caller included complete data in the UserInfo object.

Reply:
Ok - I think I have a better understanding after reading some of the other areas of the code. Let me rephrase here to see if I have a good understanding:

1. We use the usename/principalid as part of the cluster role that we create for the impersonation logic. So we have to extract those values because we need them in the `rulesForUser()` call, but we don't need any other extras, so we don't bother extracting those.
2. This new way of getting extras (from the attributes rather than from the user info) is the preferred way and should be used from now on. I am assuming that we will eventually get around to switching the other areas that use this code at some point in the future.
3. We expect the userInfo to have a subset of what's in the userAttributes. Anything more would be a separate defect and not part of this PR's scope.

Let me know if any of this is incorrect. But I think that this makes sense to me now.

Reply:
> but we don't need any other extras, so we don't bother extracting those.

There are no other extras, there are only these two.

> This new way of getting extras (from the attributes rather than from the user info) is the preferred way and should be used from now on.

It would depend on what it is being used for and what is creating the user info object. The user info object isn't a static state stored in etcd, it's an object that any component can create on the fly.

> We expect the userInfo to have a subset of what's in the userAttributes.

With this change, the only requirement for the user info object is that it has the user's UID. That is used to look up the User and UserAttributes.

Reply:
> There are no other extras, there are only these two.

Correct me if I am wrong, but isn't the extras just arbitrary key(string), value (string) pairs that the auth provider thought was important? My source was the [k8s docs](https://kubernetes.io/docs/reference/access-authn-authz/authentication/), but I might be misreading what is there. I would think that this could include any number of values outside of just principalid and username.

> It would depend on what it is being used for and what is creating the user info object. The user info object isn't a static state stored in etcd, it's an object that any component can create on the fly.

Thanks for explaining. I think that I understand. I'm thinking in light of this PR - we would want the ultimate source of the extras (anywhere else) to be the user attributes. But that's out of scope here - I don't mean to get us off track.

Reply:
> Correct me if I am wrong, but isn't the extras just arbitrary key(string), value (string) pairs that the auth provider thought was important?

It is not arbitrary for Rancher, it is always principal ID and username, examples:

https://github.com/rancher/rancher/blob/c2d8e324552a9287a726e0cb7df7fe5971169e9e/pkg/auth/providers/github/github_provider.go#L412-L421
https://github.com/rancher/rancher/blob/c2d8e324552a9287a726e0cb7df7fe5971169e9e/pkg/auth/providers/saml/saml_provider.go#L399-L408
https://github.com/rancher/rancher/blob/c2d8e324552a9287a726e0cb7df7fe5971169e9e/pkg/auth/providers/ldap/ldap_provider.go#L373-L382

Remember that these are not arbitrary headers coming from the user, these are requests that are proxied through Rancher so Rancher controls the headers.